src/Model/DriverOptions.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+namespace MongoDB\Model;
+
+use Composer\InstalledVersions;
+use MongoDB\Builder\BuilderEncoder;
+use MongoDB\Codec\Encoder;
+use MongoDB\Exception\InvalidArgumentException;
+use Throwable;
+
+use function array_diff_key;
+use function array_filter;
+use function is_array;
+use function is_string;
+use function sprintf;
+use function trim;
+
+/** @internal */
+final class DriverOptions
+{
+    private const KEY_TYPE_MAP = 'typeMap';
+    private const KEY_BUILDER_ENCODER = 'builderEncoder';
+    private const KEY_AUTO_ENCRYPTION = 'autoEncryption';
+    private const KEY_DRIVER = 'driver';
+    private const DEFAULT_TYPE_MAP = [
+        'array' => BSONArray::class,
+        'document' => BSONDocument::class,
+        'root' => BSONDocument::class,
+    ];
+
+    private const HANDSHAKE_SEPARATOR = '/';
+
+    private static ?string $version = null;
+
+    private function __construct(
+        public readonly array $typeMap,
+        public readonly Encoder $builderEncoder,
+        public readonly ?array $autoEncryption,
+        public readonly array $driver,
+        private readonly array $miscOptions,
+    ) {
+    }
+
+    public static function fromArray(array $options): self
+    {
+        $options += ['typeMap' => self::DEFAULT_TYPE_MAP];
```

Comment:
Should this be `self::KEY_TYPE_MAP`?
```suggestion
        $options += ['self::KEY_TYPE_MAP' => self::DEFAULT_TYPE_MAP];
```